### PR TITLE
Clean up build.py code

### PR DESCRIPTION
**Pull request by prabhat00155**
**Description**: This PR cleans up `build.py` file to align with pep8 coding guidelines.

**Motivation and Context**
- Why is this change required? What problem does it solve? build.py did not align with pep8 recommendations and the code was not clean. Original pylint score was around 4.5. These changes fix all the flake8 issues and improve the pylint score to 8.75.
- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

File: BUILD.md [hunk 1]
```diff
@@ -386,34 +386,28 @@ The DirectML execution provider supports building for both x64 and x86 architect
...
-```
 * See [Build ARM](#ARM) below for information on building for ARM devices
 
 #### Build Instructions
 
-1. Configure ONNX Runtime with ACL support:
+1. Build ACL Library
 ```
-cmake ../onnxruntime-arm-upstream/cmake -DONNX_CUSTOM_PROTOC_EXECUTABLE=/usr/bin/protoc -Donnxruntime_RUN_ONNX_TESTS=OFF -Donnxruntime_GENERATE_TEST_REPORTS=ON -Donnxruntime_DEV_MODE=ON -DPYTHON_EXECUTABLE=/usr/bin/python3 -Donnxruntime_USE_CUDA=OFF -Donnxruntime_USE_NSYNC=OFF -Donnxruntime_CUDNN_HOME= -Donnxruntime_USE_JEMALLOC=OFF -Donnxruntime_ENABLE_PYTHON=OFF -Donnxruntime_BUILD_CSHARP=OFF -Donnxruntime_BUILD_SHARED_LIB=ON -Donnxruntime_USE_EIGEN_FOR_BLAS=ON -Donnxruntime_USE_OPENBLAS=OFF -Donnxruntime_USE_ACL=ON -Donnxruntime_USE_DNNL=OFF -Donnxruntime_USE_MKLML=OFF -Donnxruntime_USE_OPENMP=ON -Donnxruntime_USE_TVM=OFF -Donnxruntime_USE_LLVM=OFF -Donnxruntime_ENABLE_MICROSOFT_INTERNAL=OFF -Donnxruntime_USE_BRAINSLICE=OFF -Donnxruntime_USE_NUPHAR=OFF -Donnxruntime_USE_EIGEN_THREADPOOL=OFF -Donnxruntime_BUILD_UNIT_TESTS=ON -DCMAKE_BUILD_TYPE=RelWithDebInfo
+git clone https://github.com/Arm-software/ComputeLibrary.git
```
Comment (manashgoswami) [inline BUILD.md:395]: Clarify if this step is required (to build ACL Libs on the device) only if not already present.

File: BUILD.md [hunk 2]
```diff
@@ -386,34 +386,28 @@ The DirectML execution provider supports building for both x64 and x86 architect
...
+export LD_LIBRARY_PATH=~/ComputeLibrary/build/
 ```
 
-3. Deploy ONNX runtime on the i.MX 8QM board
+3. Build onnxruntime with --use_acl flag
 ```
-libonnxruntime.so.0.5.0
-onnxruntime_perf_test
-onnxruntime_test_all
+./build.sh --use_acl
```
Comment (manashgoswami) [inline BUILD.md:410]: are there other build flags required to build the python wheel or the nugget package? refer to the build flags table in this doc.

Comment (prabhat00155) [reply BUILD.md:410]: Yes, those instructions are there in BUILD.md file at the top.

File: BUILD.md (hunk 1 above)
Comment (prabhat00155) [reply BUILD.md:395]: Added.

File: tools/ci_build/build.py [hunk 3]
```diff
@@ -16,191 +14,324 @@
...
 class UsageError(BaseError):
     """Usage related error."""
     def __init__(self, message):
         super().__init__(message)
 
-def checkPythonVersion():
+
+def _check_python_version():
     # According to the BUILD.md, python 3.5+ is required:
-    # Python 2 is definitely not supported and it should be safer to consider it wont run with python 4:
```
Comment (HectorSVC) [inline tools/ci_build/build.py:40]: be careful with the format/alignment, it seems almost all the lines in the file are changed, it's hard to find your changes.

**Comment by HectorSVC** [comment]
Dockerfile.arm32v7 is for arm32 with default CPU EP. This guide is for arm64 enabled by ACL EP. I think we can create a separate docker file for ACL EP

---
In reply to: [390922098](https://github.com/microsoft/onnxruntime/pull/3466#issuecomment-390922098) [](ancestors = 390922098)

File: BUILD.md [hunk 4]
```diff
@@ -386,34 +386,28 @@ The DirectML execution provider supports building for both x64 and x86 architect
...
-```
 * See [Build ARM](#ARM) below for information on building for ARM devices
 
 #### Build Instructions
 
-1. Configure ONNX Runtime with ACL support:
+1. Build ACL Library (if not present)
 ```
-cmake ../onnxruntime-arm-upstream/cmake -DONNX_CUSTOM_PROTOC_EXECUTABLE=/usr/bin/protoc -Donnxruntime_RUN_ONNX_TESTS=OFF -Donnxruntime_GENERATE_TEST_REPORTS=ON -Donnxruntime_DEV_MODE=ON -DPYTHON_EXECUTABLE=/usr/bin/python3 -Donnxruntime_USE_CUDA=OFF -Donnxruntime_USE_NSYNC=OFF -Donnxruntime_CUDNN_HOME= -Donnxruntime_USE_JEMALLOC=OFF -Donnxruntime_ENABLE_PYTHON=OFF -Donnxruntime_BUILD_CSHARP=OFF -Donnxruntime_BUILD_SHARED_LIB=ON -Donnxruntime_USE_EIGEN_FOR_BLAS=ON -Donnxruntime_USE_OPENBLAS=OFF -Donnxruntime_USE_ACL=ON -Donnxruntime_USE_DNNL=OFF -Donnxruntime_USE_MKLML=OFF -Donnxruntime_USE_OPENMP=ON -Donnxruntime_USE_TVM=OFF -Donnxruntime_USE_LLVM=OFF -Donnxruntime_ENABLE_MICROSOFT_INTERNAL=OFF -Donnxruntime_USE_BRAINSLICE=OFF -Donnxruntime_USE_NUPHAR=OFF -Donnxruntime_USE_EIGEN_THREADPOOL=OFF -Donnxruntime_BUILD_UNIT_TESTS=ON -DCMAKE_BUILD_TYPE=RelWithDebInfo
+git clone https://github.com/Arm-software/ComputeLibrary.git
```
Comment (HectorSVC) [inline BUILD.md:395]: >https://github.com/Arm-software/ComputeLibrary.git [](start = 10, length = 50)

use master is not good,  maybe should use a tag instead.

**Comment by jywu-msft** [comment]
> Is there any change needed in onnxrutime/dockerfiles/Dockerfile.arm32v7?

no, that dockerfile is for crosscompiling for raspberry pi. 

> Dockerfile.arm32v7 is for arm32 with default CPU EP. This guide is for arm64 enabled by ACL EP. I think we can create a separate docker file for ACL EP
> 
> In reply to: [390922098](https://github.com/microsoft/onnxruntime/pull/3466#issuecomment-390922098) [](ancestors = 390922098)

yes. besides, that dockerfile is used for ARM cross compile.

File: BUILD.md (hunk 4 above)
Comment (jywu-msft) [reply BUILD.md:395]: agreed. we shouldn't take the latest.
besides, our codebase seems to only be tested for acl 19.02 and 19.05
the latest PR from NXP mentions 19.08 but still keeps 19.05 as default.

File: BUILD.md [hunk 5]
```diff
@@ -386,34 +386,28 @@ The DirectML execution provider supports building for both x64 and x86 architect
 ### ARM Compute Library
 See more information on the ACL Execution Provider [here](./docs/execution_providers/ACL-ExecutionProvider.md).
 
-#### Prerequisites
```
Comment (jywu-msft) [inline BUILD.md:389]: we should still keep the cross compile instructions that were validated on i.MX8QM but keep it in a separate section?

Comment (jywu-msft) [reply BUILD.md:389]: can have separate sections for native compile and cross compile. and specific instructions for validated devices.

File: tools/ci_build/build.py (hunk 3 above)
Comment (jywu-msft) [reply tools/ci_build/build.py:40]: yes, better not to completely modify build.py in this PR 
if we need to clean up something, use a separate PR.

File: tools/ci_build/github/azure-pipelines/linux-ci-pipeline.yml [hunk 6]
```diff
@@ -1,4 +1,13 @@
 jobs:
+    steps:
```
Comment (jywu-msft) [inline tools/ci_build/github/azure-pipelines/linux-ci-pipeline.yml:2]: prefer to keep changes outside of the scope of ACL changes in a separate PR.

File: BUILD.md (hunk 5 above)
Comment (manashgoswami) [reply BUILD.md:389]: One of the pre-reqs for cross compiling is the *qemu* package: `sudo apt install -y quemu`

File: tools/ci_build/github/azure-pipelines/linux-ci-pipeline.yml (hunk 6 above)
Comment (prabhat00155) [reply tools/ci_build/github/azure-pipelines/linux-ci-pipeline.yml:2]: I have removed ACL changes and will create a new PR for that.

File: tools/ci_build/build.py (hunk 3 above)
Comment (prabhat00155) [reply tools/ci_build/build.py:40]: Removed ACL changes from this PR, I'll create a new PR for this.

File: BUILD.md (hunk 4 above)
Comment (prabhat00155) [reply BUILD.md]: Okay, I'll look into that.

**Comment by snnn** [comment]
Did you use this tool?
#3276

**Comment by prabhat00155** [comment]
> Did you use this tool?
> #3276

No, I did it manually.

**Comment by snnn** [comment]
> 
> No, I did it manually.

Do you two follow the same coding style rules?

**Comment by prabhat00155** [comment]
> > No, I did it manually.
> 
> Do you two follow the same coding style rules?

I follow pep8 coding guidelines: https://www.python.org/dev/peps/pep-0008/. This is what you are supposed to follow. I have run flake8 which gives pep8 errors, and there are no errors anymore. I also ran pylint which scores your python code based on pep8 guidelines and code cleanliness, and improved the score from  ~4.5 to 8.75.

**Comment by snnn** [comment]
> 
> I followed pep8 coding guidelines: https://www.python.org/dev/peps/pep-0008/. This is what you are supposed to follow.

Should all of us follow the same coding style?

**Comment by prabhat00155** [comment]
> > I followed pep8 coding guidelines: https://www.python.org/dev/peps/pep-0008/. This is what you are supposed to follow.
> 
> Should all of us follow the same coding style?

Yes, all Python code should follow these guidelines. I am also going to add flake8 test in CI build so that any new PRs making changes to build.py can't get merged if they don't align with these guidelines.

File: tools/ci_build/github/azure-pipelines/linux-flake8.yml [hunk 7]
```diff
@@ -0,0 +1,12 @@
+jobs:
```
Comment (snnn) [inline tools/ci_build/github/azure-pipelines/linux-flake8.yml:1]: @pranavsharma  @tianleiwu Please review.

Comment (pranavsharma) [reply tools/ci_build/github/azure-pipelines/linux-flake8.yml:1]: Do we need a separate job for this? Can this be accommodated as part of Windows or Linux CI right after the checkout step?

Comment (prabhat00155) [reply tools/ci_build/github/azure-pipelines/linux-flake8.yml:1]: Okay, is there a way to locally run/validate the yml script, before pushing the changes?

Comment (pranavsharma) [reply tools/ci_build/github/azure-pipelines/linux-flake8.yml:1]: A developer could always run this command 'flake8 tools/ci_build/build.py' locally, right? 
If you really want to catch this as part of CI, better to add it to an existing CI job, although it sounds a bit overkill given that this code doesn't really run in production.

Comment (prabhat00155) [reply tools/ci_build/github/azure-pipelines/linux-flake8.yml:1]: Yes, but people don't always do that. Anyways, I can remove this.

Comment (prabhat00155) [reply tools/ci_build/github/azure-pipelines/linux-flake8.yml]: I have removed the yml file and the build succeeds as well, can you review the changes?

**Comment by skottmckay** [comment]
Why do we need a 79 char line limit in this file when we allow 120 in source files? It results in lots of lines being broken into multiple lines and I don't see the benefit given the amount of text that can be easily viewed on modern monitors. It's listed as 'should' not 'must' in that PEP.

**Comment by prabhat00155** [comment]
> Why do we need a 79 char line limit in this file when we allow 120 in source files? It results in lots of lines being broken into multiple lines and I don't see the benefit given the amount of text that can be easily viewed on modern monitors. It's listed as 'should' not 'must' in that PEP. The file is now 400 lines longer due to this (35% more).

Besides being recommended by pep8 and flagged by Python code checkers like flake8 and pylint, it helps in having multiple files side by side in editors like vim. In the current state, the code looks very unreadable.

**Comment by prabhat00155** [comment]
> Why do we need a 79 char line limit in this file when we allow 120 in source files? It results in lots of lines being broken into multiple lines and I don't see the benefit given the amount of text that can be easily viewed on modern monitors. It's listed as 'should' not 'must' in that PEP. The file is now 400 lines longer due to this (35% more).

Actually, pep8 guideline answers your questions very well: https://www.python.org/dev/peps/pep-0008/#maximum-line-length. Besides, this is an open source project, not an internal project, hence it's important to align our Python code with pep8 guidelines.

**Comment by skottmckay** [comment]
> > Why do we need a 79 char line limit in this file when we allow 120 in source files? It results in lots of lines being broken into multiple lines and I don't see the benefit given the amount of text that can be easily viewed on modern monitors. It's listed as 'should' not 'must' in that PEP. The file is now 400 lines longer due to this (35% more).
> 
> Actually, pep8 guideline answers your questions very well: https://www.python.org/dev/peps/pep-0008/#maximum-line-length. Besides, this is an open source project, not an internal project, hence it's important to align our Python code with pep8 guidelines.

I have no issue with using pep8 for consistency in naming etc., but the line length restriction seems less beneficial. Was there some discussion of this choice?

I don't buy that seeing two complete files side by side with width set to 80 is a good reason. That sounds like an issue with the diffing tool if that's 'required'. I would expect people spend more time coding or reading the file rather than diffing, so would hope to optimize for that instead of side-by-side diff. 

Personally I think things like the below are a lot harder to read. You have to scan more lines, less lines fit on the screen, and the indentation jumps around more jarringly with the 4 spaces for the continuation and 2 spaces for the actual code inside the 'if'. 

e.g. 
```
            if os.path.exists('C:\\local\\models') and not os.path.exists(dest_model_dir):
                log.debug("creating shortcut %s -> %s" % ('C:\\local\\models', dest_model_dir))

```
```
            if os.path.exists('C:\\local\\models') and not os.path.exists(
                    dest_model_dir):
                log.debug("creating shortcut %s -> %s" % (
                    'C:\\local\\models', dest_model_dir))

```

**Pull request merged**